src/permissions_template.c

```diff
@@ -0,0 +1,117 @@
+/*
+Copyright (C) 2015 The University of Notre Dame
+This software is distributed under the GNU General Public License.
+See the file LICENSE for details.
+*/
+
+
+#include "kerneltypes.h"
+#include "kmalloc.h"
+#include "permissions_template.h"
+#include "console.h"
+
+#define MAX_NUMBER_OF_PERMISSIONS 169 // temporary, for now they'll be statically allocated
+#define DEFAULT_FRACTION_OF_MEMORY 3
+
+static uint32_t template_counter = 1;
+static struct permissions_template* permissions_table[MAX_NUMBER_OF_PERMISSIONS]; // statically allocated array of templates
+
+uint32_t generate_new_identifier() {
+    return template_counter++;
+}
+
+struct process_permissions* permissions_from_template(struct permissions_template *template) {
+
+    if (template == 0) {
+        console_printf("Error: invalid template\n");
+        return 0; // error case, incomplete permissions. maybe use default permissions?
+    }
+    struct process_permissions *new_permissions = kmalloc(sizeof(struct process_permissions));
+    if (new_permissions == 0) {
+        console_printf("Error: could not allocate for permissions\n");
+        return new_permissions;
+    }
+    // apply information
+    new_permissions->max_number_of_pages = template->max_number_of_pages;
+    new_permissions->max_width = template->max_width;
+    new_permissions->max_height = template->max_height;
+    new_permissions->offset_x = template->offset_x;
+    new_permissions->offset_y = template->offset_y;
+
+    // todo: the rest of the permissions
```

Comment:
We'll have to be careful with the fs permissions here in the future, as it looks like we'll need to copy a linked list as opposed to just copying a pointer (the parent could then delete the template!).